### ISSUE TITLE
Wobbly highlighting of menu entries in XaAES

### ISSUE THREAD

**Issue by GokMasE** [freemint/freemint#248]
When it comes to visual feedback in menus, XaAES will highlight the entry that mouse is currently hovering over - this looks decent as long as the menu is a standard menu with just entries / separators. If the mouse is moved on top of a separator (or any other disabled menu entry), then highlighting is disabled/turned off until next time mouse is moved on top of a non-disabled entry. All cool.

However, if a menu has a combination of "normal entries", "separators" and "entries with submenus attached", things get a bit messy.
Moving the mouse up/down such a menu after having triggered a submenu to open, will (depending on the speed the mouse is moved with) cause the highglighting to "jump" between the entry currently beneath the mouse and the menu entry that was responsible for opening the submenu.

**Why is that? Well, when mouse moves on top of a separator XaAES don't want to highglight it (naturally, since it can't be selected) but instead of just highlighting nothing XaAES then decides that if there is a submenu still open (which has not yet timed out) it will highlight the menu entry that was responsible for opening it.  The result can look really dodgy and flickering which contributes to a rather strange and wobbly look and feel.**

Another oddity, is that XaAES will keep any submenu not yet timed out (and thus still open) open forever as long as the mouse is on top of a separator - it will never timeout.

Here is an [illustrating video](http://xaaes.gokmase.com/wobbly_menu_bug.mp4) that aims to show how the highlighting "sneaks back" to the entry linked to the submenu each time the mouse hits the separator/disabled entry. I apologize for the crappy quality, hopefully it will still give an idea.

Some thoughts:

- Maybe it would make sense to simply highlight nothing at all whenever mouse is on top of a separator? (regardless of any open popup menus)

- It might also make sense to include the separators when keeping track of the popout_timeout value to decide when a popup menu should be closed - atm locating the mouse on top of a separator keeps any submenu open forever, instead of (as would perhaps be expected) closing it after popout_timeout value has been reached.

- Attention probably needs to be paid to making sure any changes are playing well with the keyboard navigation (which can trigger menu entry highglighting too).

**NOTE: The highlighting behaviour explained above seems to be the exact same on both XaAES-current and XaAES-ozk.**

**Comment by mikrosk** [comment]
I'm pretty sure that it is the timeout which makes things looking so broken. Imagine (maybe it even works) if timeout = 0, i.e. as soon as you leave the submenu entry, the submenu should close immediately. So no problem, you are on a separator, the submenu is closed, no need to jump anywhere.

I wonder what should be the correct behaviour with keyboard: I'm on a submenu entry. I move up with mouse to a separator. Then it's clear, the mouse cursor should point to the separator and the highlighting disappear. However if I do the same with keyboard... should the highlighting really disappear? Shouldn't the highlighted line jump straight to the first non-separator entry? (or not jump at all)

**Comment by GokMasE** [comment]
Yes, the timeout value makes the problem stand out a lot more than it otherwise would, so setting the popout_timeout = 0 might hide the problem while it won't (really) solve it.

I will do some tests tonight and see if popout_timeout = 0 will work the way we expect in this regard.

Navigating menus with the keyboard, I think you first need to "fold in"/close the submenu before you can navigate up/down with the keyboard in the main/parental menu. Navigating with the keyboard, the highlighting is really the only way to indicate the current focus, so I think the highlighting should never disappear in that context, it should probably just jump straight to the next selectable (that is, not a separator or "normal" but disabled entry) menu entry.

**Comment by mikrosk** [comment]
>  I think you first need to "fold in"/close the submenu before you can navigate up/down with the keyboard

So essentially, the timeout should be ignored in such case? I don't see a point where you would navigate to a submenu entry with keyboard and it would close automatically. Actually, what happens if you do that with mouse? Is that timeout applied even if you are on the submenu entry all the time?

**Comment by GokMasE** [comment]
The timeout comes into play only _after_ the mouse has left the submenu (or the main menu entry that triggered it to open).

And yes, I absolutely think that closing a submenu through keyboard input should be instantaneous, ie. not involve a timeout.
(I can't test that now, but it probably works that way already.)

**Comment by mikrosk** [comment]
I see, thanks for clarification. This should make the bug much easier to fix. I have personally hated that behaviour quite a lot. ;-)

**Comment by GokMasE** [comment]
I can confirm that (as predicted) there is no timeout value involved when "folding in" an open submenu by pressing left arrow button, the submenu is currently closed immediately regardless of the popout_timeout value when keyboard is triggering the close.

I can also confirm that (exactly as we suspected) setting popout_timeout = 0 indeed seems to fully hide the wobbly pattern of the highlighting that can otherwise take place in menus that has both non-selectable entries (separators or disabled menu entries), normal entries and entries with submenus attached. I can conclude that it is a great "quick fix" to hide the bad behaviour, but not a cure for the actual bug :)
